hazelcast/src/main/java/com/hazelcast/map/impl/InterceptorRegistry.java

```diff
@@ -0,0 +1,122 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl;
+
+import com.hazelcast.map.MapInterceptor;
+import com.hazelcast.spi.impl.operationexecutor.classic.PartitionOperationThread;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.EMPTY_LIST;
+import static java.util.Collections.EMPTY_MAP;
+
+/**
+ * Registry for all {@link MapInterceptor} of an {@code IMap}
+ *
+ * Interceptors are read mostly and this registry keeps all registered interceptor in an array to easily iterate on them.
+ * Other than that, synchronized blocks are used to prevent leaks when concurrently registering/de-registering interceptors.
+ * Keep in mind that all registration/de-registration operations are done in generic-operation-threads, in other words,
+ * synchronized methods are not used in partition-threads.
+ *
+ * This registry is created per map.
+ */
+public class InterceptorRegistry {
+
+    private volatile List<MapInterceptor> interceptors = EMPTY_LIST;
+    private volatile Map<String, MapInterceptor> id2InterceptorMap = EMPTY_MAP;
+
+    /**
+     * Gives an instance of this registry according to generic-operation-thread count.
+     */
+    public InterceptorRegistry() {
+    }
+
+    /**
+     * Returns all registered interceptors.
+     *
+     * This method is called by {@link PartitionOperationThread}
+     *
+     * @return all registered interceptors
+     * @see com.hazelcast.map.impl.recordstore.DefaultRecordStore#put
+     */
+    public List<MapInterceptor> getInterceptors() {
+        return interceptors;
+    }
+
+    public Map<String, MapInterceptor> getId2InterceptorMap() {
+        return id2InterceptorMap;
+    }
+
+    /**
+     * Registers a {@link MapInterceptor} for the supplied `id`.
+     * If there is no registration associated with the `id`, registers interceptor,
+     * otherwise silently ignores registration.
+     *
+     * This method is called by {@link com.hazelcast.spi.impl.operationexecutor.classic.GenericOperationThread}
+     * when registering via {@link com.hazelcast.map.impl.operation.AddInterceptorOperation}
+     *
+     * @param id          id of the interceptor
+     * @param interceptor supplied {@link MapInterceptor}
+     */
+    public synchronized void register(String id, MapInterceptor interceptor) {
+        assert !(Thread.currentThread() instanceof PartitionOperationThread);
+
+        if (id2InterceptorMap.containsKey(id)) {
+            return;
+        }
+
+        Map<String, MapInterceptor> tmpMap = new HashMap<String, MapInterceptor>(id2InterceptorMap);
+        tmpMap.put(id, interceptor);
+
+        id2InterceptorMap = tmpMap;
```

Comment:
Can you also wrap with `Collections.unmodifiableMap()`? Readers should not be able to modify.

```
id2InterceptorMap = Collections.unmodifiableMap(tmpMap);
```